File: backend/routers/query.py
from fastapi import APIRouter, HTTPException

from models.schemas import Citation, QueryRequest, QueryResponse
from services.embedder import get_query_embedding
from services.llm import generate_answer
from services.vector_store import search_similar

router = APIRouter(tags=["query"])


@router.post("/query", response_model=QueryResponse, response_model_by_alias=True)
async def query_documents(request: QueryRequest):
    if not request.question.strip():
        raise HTTPException(status_code=400, detail="Question cannot be empty.")

    try:
        query_embedding = get_query_embedding(request.question)
        similar_chunks = search_similar(query_embedding, top_k=request.top_k)

        if not similar_chunks:
            return QueryResponse(
                answer="No relevant documents found. Please upload documents before querying.",
                citations=[],
            )

        answer = generate_answer(request.question, similar_chunks)

        # Deduplicate citations by (doc, page), preserve score
        citations: list[Citation] = []
        seen: set[tuple] = set()
        for chunk in similar_chunks:
            key = (chunk["document_name"], chunk["page_number"])
            if key not in seen:
                preview = chunk["text"][:200] + "…" if len(chunk["text"]) > 200 else chunk["text"]
                citations.append(
                    Citation(
                        document_name=chunk["document_name"],
                        page_number=chunk["page_number"],
                        chunk_text=preview,
                        score=chunk.get("score", 0.0),
                    )
                )
                seen.add(key)

        return QueryResponse(answer=answer, citations=citations)

    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc

File: server/routes/query.js
import express from 'express'
import { getQueryEmbedding } from '../services/embedder.js'
import { searchSimilar } from '../services/vectorStore.js'
import { generateAnswer } from '../services/llm.js'

const router = express.Router()

// POST /api/query
router.post('/query', async (req, res) => {
  const { question } = req.body
  const top_k = Math.min(Math.max(parseInt(req.body.top_k, 10) || 5, 1), 20)

  if (!question?.trim()) {
    return res.status(400).json({ error: 'Question cannot be empty.' })
  }

  if (question.length > 2000) {
    return res.status(400).json({ error: 'Question is too long (max 2000 characters).' })
  }

  try {
    // 1. Embed the query
    const queryEmbedding = await getQueryEmbedding(question.trim())

    // 2. Retrieve top-k similar chunks scoped to this user
    const similarChunks = searchSimilar(queryEmbedding, req.user.uid, top_k)

    if (!similarChunks.length) {
      return res.json({
        answer: 'No relevant documents found. Please upload documents before querying.',
        citations: [],
      })
    }

    // 3. Generate grounded answer
    const answer = await generateAnswer(question.trim(), similarChunks)

    // 4. Deduplicate citations by (doc, page)
    const seen = new Set()
    const citations = []
    for (const chunk of similarChunks) {
      const key = `${chunk.documentName}::${chunk.pageNumber}`
      if (!seen.has(key)) {
        seen.add(key)
        citations.push({
          documentName: chunk.documentName,
          pageNumber:   chunk.pageNumber,
          chunkText:    chunk.text.length > 250 ? chunk.text.slice(0, 250) + '…' : chunk.text,
          score:        Math.round(chunk.score * 100) / 100,
        })
      }
    }

    res.json({ answer, citations })
  } catch (err) {
    console.error('[Query Error]', err)
    const msg = err.message?.toLowerCase() || ''
    const isQuota = err.status === 429 || msg.includes('quota') || msg.includes('rate limit') || msg.includes('429')
    res.status(500).json({
      error: isQuota
        ? 'AI service quota exceeded. Please try again later.'
        : 'Failed to process query.',
    })
  }
})

export default router
